### PR TITLE
Replace already closed github.com/golang/mock with go.uber.org/mock

### LINKED ISSUE #2356: Replace already closed github.com/golang/mock with go.uber.org/mock
### What you would like to be added?

I would like to replace the already closed github.com/golang/mock with the new go.uber.org/mock as recommended in the following:

> Update, June 2023: This repo and tool are no longer maintained. Please see [go.uber.org/mock](https://github.com/uber/mock) for a maintained fork instead.

https://github.com/golang/mock?tab=readme-ov-file#gomock

Additionally, we can improve the order dependencies integration tests using [WithOverridableExpectations](https://pkg.go.dev/go.uber.org/mock@v0.4.0/gomock#WithOverridableExpectations) as we discussed here: https://github.com/kubeflow/katib/pull/2350#discussion_r1638688994

### Why is this needed?

The github.com/golang/mock is no longer maintained.

### Love this feature?

Give it a 👍 We prioritize the features with most 👍

### PR DESCRIPTION
**Pull request by forsaken628**
**What this PR does / why we need it**:

github.com/golang/mock already closed 

**Which issue(s) this PR fixes** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_:
Fixes #2356

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/katib/) included if any changes are user facing

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -188,3 +183,31 @@ pytest-skopt:
 	pip install --prefer-binary -r test/unit/v1beta1/requirements.txt
 	pip install --prefer-binary -r cmd/suggestion/skopt/v1beta1/requirements.txt
 	PYTHONPATH=$(PYTHONPATH) pytest ./test/unit/v1beta1/suggestion/test_skopt_service.py
+
+##@ Dependencies
```
Comment (tenzen-y) [inline Makefile:187]: Could you remove the irrelevant commit from this PR?

File: Makefile [hunk 2]
```diff
@@ -188,3 +183,31 @@ pytest-skopt:
...
+	mkdir -p $(LOCALBIN)
+
+MOCKGEN = $(LOCALBIN)/mockgen-$(MOCKGEN_VERSION)
+
+.PHONY: mockgen
+mockgen: $(MOCKGEN) ## Download mockgen locally if necessary.
+$(MOCKGEN): $(LOCALBIN)
+	$(call go-install-tool,$(MOCKGEN),go.uber.org/mock/mockgen,${MOCKGEN_VERSION})
+
+# go-install-tool will 'go install' any package with custom target and name of binary, if it doesn't exist
```
Comment (tenzen-y) [inline Makefile:201]: Could you remove the irrelevant commit from this PR?

**Comment by tenzen-y** [comment]
@forsaken628 One question here, do we need to introduce https://github.com/uber/mock/blob/v0.4.0/gomock/controller.go#L116 the as we discussed here?: https://github.com/kubeflow/katib/pull/2350#discussion_r1638688994

**Comment by forsaken628** [comment]
It's a bit more rigorous when it's introduced, always gomock.InOrder makes it work too.

**Comment by tenzen-y** [comment]
@forsaken628 It seems that the e2e tests often failed due to webhook server startup.
I was supposed to add the custom readiness check in the https://github.com/kubeflow/katib/pull/2345 like the training-operator: https://github.com/kubeflow/training-operator/blob/9a9edecf146bb0c2cccaeb3bec912ab2dfa5d8c5/cmd/training-operator.v1/main.go#L230-L247

So, could you open PR to implement such a custom readiness chack mechanism to the katib controller?

**Comment by tenzen-y** [comment]
> It's a bit more rigorous when `WithOverridableExpectations` is introduced, always `gomock.InOrder` makes it work too.

Uhm, that makes sense.
Let's revisit here once we need to add the more strict verifications since the current implementations align with the previous implementations.

**Comment by tenzen-y** [comment]
Anyway, this PR still lgtm
/lgtm
/approve

**Pull request merged**